package/MDAnalysis/lib/formats/libmdaxdr.pyx

```diff
@@ -747,6 +892,67 @@ cdef class XTCFile(_XDRFile):
             self.current_frame += 1
         return XTCFrame(xyz, box, step, time, prec)
 
+    def read_direct(self, np.float32_t[:, ::1] positions):
+        """
+        Read next frame in the XTC file with positions read directly into
+        a pre-existing array.
+
+        Parameters
+        ----------
+        positions : np.ndarray
+           positions array to read positions into
+
+        Returns
+        -------
+        frame : libmdaxdr.XTCFrame
+            namedtuple with frame information
+
+        See Also
+        --------
+        XTCFrame
+        TRRFile
+
+        Raises
+        ------
+        IOError
+
+
+        .. versionadded:: 2.4.0
+        """
+        if self.reached_eof:
+            raise EOFError('Reached last frame in XTC, seek to 0')
+        if not self.is_open:
+            raise IOError('No file opened')
+        if self.mode != 'r':
+            raise IOError('File opened in mode: {}. Reading only allow '
+                               'in mode "r"'.format('self.mode'))
+
+        return_code = 1
+        cdef int step
+        cdef float time, prec
+        cdef np.npy_intp[2] unitcell_dim
+        unitcell_dim[0] = DIMS
+        unitcell_dim[1] = DIMS
+
+        cdef np.ndarray[np.float32_t, ndim=2] box = np.PyArray_EMPTY(2, unitcell_dim, np.NPY_FLOAT32, 0)
+
+
+        return_code = read_xtc(self.xfp, self.n_atoms, <int*> &step,
+                                      &time, <matrix>box.data,
+                                      <rvec*>&positions[0,0], <float*> &prec)
+        if return_code != EOK and return_code != EENDOFFILE:
+            raise IOError('XTC read error = {}'.format(
+                error_message[return_code]))
+
+        if return_code == EENDOFFILE:
+            self.reached_eof = True
+            raise StopIteration
+
+        if return_code == EOK:
+            self.current_frame += 1
```

Comment:
I think its because the `return code` enum has other options that are not nessecarily errors but indicate some kind of file state. 

```cython
cdef enum:
    EOK = 0
    EHEADER = 1
    ESTRING = 2
    EDOUBLE = 3
    EINTEGER = 4
    EFLOAT = 5
    EUNSIGNEDINTEGER = 6
    ECOMPRESSION = 7
    ECLOSE = 8
    EMAGIC = 9
    EMEMORY = 10
    EENDOFFILE = 11
    ENOTFOUND = 12

```

Reply:
So checking exclusively for EOK may break this ? Im not super sure (query @kain88-de @jbarnoud?)

Reply:
But we don't even return the state so it looks as if our reader may end up in a situation where we read a frame back but didn't increment the frame number but for the user still everything looks just fine. 

I suggest we fail harder here and wait for feedback from our tests and the community... I consider this fixing a behavior that can lead to silent errors.